open-apparel-registry/open-apparel-registry
# Add Nonstandard Fields Model

## Overview

Our previous method of looping through a contributor's facility list
items and parsing them for the contributor's nonstandard fields was
timing out on staging. This adds a NonstandardField model, which gets
created when a user creates a facility via API or creates a list,
allowing us to access the user's nonstandard fields more quickly.

Connects #1327 

## Demo

<img width="227" alt="Screen Shot 2021-05-17 at 9 08 05 AM" src="https://user-images.githubusercontent.com/21046714/118514085-4297e300-b702-11eb-9951-f066199399f9.png">

## Testing Instructions

* Run `vagrant ssh` and `./scripts/update`
* Run`./scripts/test`
     - [x] The tests should pass
* Run `./scripts/server`
* Navigate to http://localhost:8081/api/nonstandard-fields/
     - [x] You should see the default fields: parent_company, type_of_product, number_of_workers, type_of_facility
* Login as `c2@example.com` and [submit a list](http://localhost:6543/contribute) with an extra column
* Get an API token from [settings](http://localhost:6543/settings)
* In [the Swagger API](http://localhost:8081/api/docs/#!/facilities/facilities_create), authorize and then submit a facility with a different extra field
* Navigate to the [nonstandard fields api](http://localhost:8081/api/nonstandard-fields/)
     - [x] You should see both extra fields listed, as well as the default fields

## Checklist
